### DISCUSSION TITLE
Genre list/grid based on standard React Admin template

### DISCUSSION THREAD

**Discussion by certuna** [navidrome/navidrome#2631] [Ideas]
The WebUI currently has no support for navigation or filtering by genre at all. It just displays (one) genre in the `Album` view. Most other players (Apple Music, Kodi, PlexAmp, Marvis, Prism, etc) have a navigate-by-genre top level option.

This issue should hopefully open the discussion what it should look like and what the functionality would be. Please feel free to contribute with any ideas, and shoot down some of my ideas below if they're unwanted.

**Placement:**
I would propose `Genres` in the `Library` section, below `Artists` and `Songs`.

**Server side:**
Genre is not exposed in the internal API, so this is not only front-end code, also some work on the Go back end.
- [ ] Add `/genre` API command and the underlying SQL query that returns the genre list
- [ ] Maybe the back end needs an intermediate table that's pre-rendered in the database, for speed?

**Genre UI**
Two options I think. We could have both, toggled like the current Albums view, but basically:

**1: Text columns view**
- [ ] multi-select boxes, triggering the appearance of an action bar above (like in the current `Songs` view), with actions `Play` (starts randomized 200-song "multi-genre radio" play queue), `Go To Albums` (goes to album view, filtered for the selected genres), `Go To Songs` (goes to Songs view, filtered for selected genres), `Add To Playlist` (self-explanatory), `Create Smart Playlist` (for the future, when we have smart playlists)
- [ ] `Genre` : clicking on it takes you to the `Albums` view, filtered for the genre
- [ ] `Years` : Year range of the songs in the genre, eg `1984-2013`, similar way it's handled in the text version of the existing `Albums` list.
- [ ] `Songs/Albums` : number of songs and albums: `48/5`for a genre with 48 songs and 5 albums
- [ ] `Time` : total duration of all songs in the genre: `4d 15h 45m 12s` or `35:48:51`
- [ ] `Plays` : cumulative plays for all tracks in the genre
- [ ] `Rating` : average of the 1-5 ratings of all the songs in the genre

Bit boring to look at, but functional.

Mobile view ("GenreSimpleList"?) should probably drop all columns except `Genre` and `Songs/Albums`? PlexAmp is a good example of a good minimalist mobile view, it shows an alphabetically ordered genres list like `Jazz - 63 albums` , tapping on it takes you to an album covers grid of albums tagged with that genre. [Prism](https://prism-music.app/) has a similar Genres list, with `Jazz - 2,345 songs`, tapping it takes you to a table of albums.

**2: Image grid view**

An image grid like the "Categories" section of the [React Admin demo website](https://marmelab.com/react-admin-demo/#/categories):

![Screen Shot 2021-03-15 at 23 11 45](https://user-images.githubusercontent.com/62144283/111230451-ede3c900-85e7-11eb-827f-489240975fb7.png)

Either sorted alphabetically, or descending by number of songs, or perhaps by total time?

Where the demo website has `Products` and `Edit` links, we'd have `Albums` and `Songs` links, taking you to the respective `Songs` and `Albums` view, with the relevant `Genre` filter applied. Similar to how currently, when you click on an artist in the `Artists` list, you're taken to the `Album` view with the relevant artist filter applied.

A `Play` button overlay on the genre image (on-mouseover in desktop, always visible in mobile) that starts a play queue with 200 random songs in that genre, similar to the current `Shuffle All` button in `Song` view. This basically gives Navidrome "genre radio" ( https://github.com/navidrome/navidrome/issues/316 )

![Screen Shot 2021-03-15 at 23 11 45](https://user-images.githubusercontent.com/62144283/111306816-fb856700-8658-11eb-9ba7-9546148650bc.png)

For this, we'd need genre images, server-side:
- generate them from the library by combining 3 album covers or artist images from the Navidrome library in that genre?
- scrape or bundle royalty-free images from somewhere?
- user-customizable genre images from a folder `/genres/hip-hop.jpg` ?

**Comment by Hackerbone** [comment]
I would like to work on this. It's UI related and a bit of functionality based (sorting for filters) I think I'll in this way learn more about React Admin as well.

**Comment by Hackerbone** [comment]
Is there a standard protocol to implement sorting of albums on the basis of genres and year basis which I should follow? @certuna . Can you give me some references to checkout?

**Comment by certuna** [comment]
`GET https://myserver.com/app/api/album?_end=18&_order=DESC&_sort=maxYear&_start=0` returns an album list sorted by Year.

At the moment Genre is usable through the Subsonic API but it's not used anywhere in the WebUI yet, so I think you'd have to ask @deluan (easiest on Discord) if genre is exposed through the internal API yet, and if so, how.

**Comment by Hackerbone** [comment]
I'll ask and start working on it. Thanks !

**Comment by KashyapPavra** [comment]
Hey I would also love to work on this, this seems a big task. @Hackerbone So how about we both work on it together?

**Comment by Hackerbone** [comment]
Yeah sure we can work together on this @KashyapPavra

**Comment by KashyapPavra** [comment]
@Hackerbone Great, did you have a talk with @deluan whether the genre is exposed in internal API?

**Comment by Hackerbone** [comment]
@KashyapPavra not yet, will do till then can you share your discord id?

**Comment by KashyapPavra** [comment]
> @KashyapPavra not yet, will do till then can you share your discord id?

kashyappavra#4305

**Comment by certuna** [comment]
From what I can see:

1. set up the underlying SQL query in https://github.com/navidrome/navidrome/blob/master/persistence/genre_repository.go . @deluan has already written a query there, returning `genre name`, `number of albums` and `number of songs`
2. add `/genre` command to the API in https://github.com/navidrome/navidrome/blob/master/server/app/app.go

**Comment by KashyapPavra** [comment]
@certuna 
I am not well aware with GO, so if you could implement the backend for it on GO. I'll add the front-end functionality for it

**Comment by certuna** [comment]
@KashyraPavra & @Hackerbone : I am very sorry to say this enhancement cannot be implemented yet. It's not trivial to expose a `/genre` REST endpoint since the database does not have a genre table yet. Overhauling the database schema is planned for later, so for now have edited the start post to put this enhancement on hold.

**Comment by deluan** [comment]
This requires more work on the backend. We need a table for `genre`, which will be implemented in #119

**Comment by SpencerDub** [comment]
> user-customizable genre images from a folder `/genres/hip-hop.jpg` ?

Just chiming in to say that I would love this option. I don't think it makes sense as a default, but a decade ago, I was enamored with iTunes' genre icons, and it would tickle me to be able to add my own custom icons for genres on my Navidrome server.

**Comment by metalheim** [comment]
I think now that genres are implemented correctly #1251 , there are no barriers to implementing this. 

Maybe first without images (as that adds other challenges), just as a list/table view similar to the "artists" list.
Similar to how Sonixd does it:

![image](https://user-images.githubusercontent.com/44159957/141454023-0100e212-6cf1-4a88-a141-32edc339bd20.png)

and then put it in the left navigation pane


![image](https://user-images.githubusercontent.com/44159957/141453721-f47bc7d7-4733-45ff-b7ff-108b6dc9d03c.png)

**Comment by andrewzah** [comment]
For the images, what about generating them every so often from the 3-5 most-played albums in that genre?

**Comment by sammypanda** [comment]
Have been looking for this feature :)

Is there any update?

**Reply by certuna** [reply]
Needs to wait until the migration to the new ReactAdmin version is done.

**Comment by mist2t** [comment]
Bump up for this feature 🙂

I think a really cool way to explore library by genre is vital.

It will be amazing once you "enter" a genre from the genres page to be able to explore further all the content related to that genre.

This means genres support for playlists, artists etc.

When you dig into "Rock" you will see everything "Rock": albums, songs, artists, rock playlists, rock faves etc.

Having a multi-genre library, genres are my "entry point" into every listening session. I'm thinking about what kind of music i'm in mood for today ... then dig further down into playlists, artists, albums etc. for that particular genre.
